### PR TITLE
PUBDEV-5977: train AutoML models on the leader node in the client mode

### PR DESCRIPTION
**Pull request by michalkurka**

### CODE REVIEW

**Comment by michalkurka** [comment]
@seb-h2o this is WIP - demonstrates the idea

**Comment by michalkurka** [comment]
I think we should move it to ModelBuilder and re-use also in GridSearch & REST API. That should cover most of the use cases? WDYT?

**Comment by seb-h2o** [comment]
@michalk: if I understand your changes correctly, it makes the client mode behave even more as a proxy as it used to, in the sense that it shouldn't do any work, although there are situations for which it was still the case before this fix?
Typically, I see that there is already logic to ensure that MRTask are executed remotely in client mode, but I can't see anything similar for more generic DTask... you guess my question now: can't we fix this in DTask then? and in `ModelBuilder#trainModel`, we would just use a DTask instead of the H2OCompleter? making also easier for future optimization when needed?

**Comment by michalkurka** [comment]
It would be perfect if we could change ModelBuilder#trainModel to do this automatically, unfortunately, that is not possible because of the transient state that ModelBuilders have. Initialized MB's are not designed to be serialized. That is why this PR creates the MB twice and sends the serializable parameters in the DTask instead of the full MB. The API stands in the way here.

I am planning to improve the API but I don't think the current API permits a generic solution to the problem.

File: h2o-core/src/main/java/hex/ModelBuilder.java [hunk 1]
```diff
@@ -234,6 +234,43 @@ private void setFinalState() {
...
+  // FIXME
+  @SuppressWarnings("unchecked")
+  public static Job<Model> trainModel(ModelBuilder builderTemplate) {
+    if (builderTemplate == null) {
+      throw new IllegalArgumentException("Builder template is not specified.");
+    }
+    if (H2O.ARGS.client) {
+      Job job = builderTemplate._job;
+      TrainModelTask tmt = new TrainModelTask(job, job._result, builderTemplate._parms);
+      H2ONode leader = H2O.CLOUD._memary[0];
```
Comment (seb-h2o) [inline h2o-core/src/main/java/hex/ModelBuilder.java]: btw, maybe we could make `H2O#leader` public

**Comment by seb-h2o** [comment]
looks good for now: would be nice to have a more systematic way to do this, through some sub class of DTask maybe? haven't been successful on my side..

File: h2o-core/src/main/java/hex/ModelBuilder.java [hunk 2]
```diff
@@ -234,6 +234,51 @@ private void setFinalState() {
     }
   }
 
+  /**
+   * Start model training using a this ModelBuilder as a template. The MB can be either used directly
+   * or if the method was invoked on a regular H2O node. If the method was called on a client node, the model builder
+   * will be used as a template only and the actual instance used for training will re-created on a remote H2O node.
+   *
+   * Warning: the nature of this method prohibits further use of this instance of the model builder after the method
+   *          is called.
```
Comment (seb-h2o) [inline h2o-core/src/main/java/hex/ModelBuilder.java]: thanks for the warning

**Pull request merged**